cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/input/BasicPartitionedFileSetInputContext.java

```diff
@@ -42,29 +46,50 @@
 
   private static final Type STRING_PARTITION_KEY_MAP_TYPE = new TypeToken<Map<String, PartitionKey>>() { }.getType();
 
-  private final Path inputPath;
   private final String mappingString;
 
-  private PartitionKey partitionKey;
+  private final List<Path> inputPaths;
+  private List<PartitionKey> partitionKeys;
 
   BasicPartitionedFileSetInputContext(MultiInputTaggedSplit multiInputTaggedSplit) {
     super(multiInputTaggedSplit.getName());
 
     InputSplit inputSplit = multiInputTaggedSplit.getInputSplit();
-    if (!(inputSplit instanceof FileSplit)) {
-      throw new IllegalArgumentException(String.format("Expected a '%s', but got '%s'.",
-                                                       FileSplit.class.getName(), inputSplit.getClass().getName()));
+
+    inputPaths = new ArrayList<>();
+
+    if (inputSplit instanceof FileSplit) {
+      inputPaths.add(((FileSplit) inputSplit).getPath());
+    } else if (inputSplit instanceof CombineFileSplit) {
+      Collections.addAll(inputPaths, ((CombineFileSplit) inputSplit).getPaths());
+    } else {
+      throw new IllegalArgumentException(String.format("Expected either a '%s' or a '%s', but got '%s'.",
+                                                       FileSplit.class.getName(), CombineFileSplit.class.getName(),
+                                                       inputSplit.getClass().getName()));
     }
-    this.inputPath = ((FileSplit) inputSplit).getPath();
+
     this.mappingString = multiInputTaggedSplit.getConf().get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
   }
 
   @Override
   public PartitionKey getInputPartitionKey() {
-    if (partitionKey == null) {
-      partitionKey = getPartitionKey(inputPath.toUri());
+    List<PartitionKey> inputPartitionKeys = getInputPartitionKeys();
+    if (inputPartitionKeys.size() != 1) {
+      throw new IllegalArgumentException(String.format("Expected only a single PartitionKey, but found: %s",
```

Comment:
Would it be better to throw a message that indicates how to fix it? Such as:
"Found multiple partition keys; use method getInputPartitionKeys() instead".